File: cpp/examples/activity_feed/activity_feed_panel/activity_feed_panel.cpp
#include "activity_feed_panel.hpp"

#include <algorithm>
#include <utility>

namespace activity_feed_example {

ActivityFeedPanel::ActivityFeedPanel() {
    setDynamicChildrenRenderer([this](int width) {
        return renderActivities(width);
    });
}

void ActivityFeedPanel::clearActivities() {
    activities_.clear();
    clearChildren();
    invalidateDynamicChildren();
}

void ActivityFeedPanel::appendActivity(std::string timestamp,
                                       std::string eventType,
                                       std::string message,
                                       bool showTimestamp) {
    activities_.push_back(ActivityItemData{
        .timestamp = std::move(timestamp),
        .eventType = std::move(eventType),
        .message = std::move(message),
        .showTimestamp = showTimestamp,
    });
    invalidateDynamicChildren();
}

void ActivityFeedPanel::setAutoScroll(bool enabled) {
    scrollView().setAutoScroll(enabled);
}

std::size_t ActivityFeedPanel::activityCount() const {
    return activities_.size();
}

int ActivityFeedPanel::viewOffset() const {
    return scrollView().viewOffset();
}

int ActivityFeedPanel::feedWidth() const {
    return std::max(1, frame().width);
}

std::vector<ui::RenderedContent> ActivityFeedPanel::renderActivities(int width) const {
    std::vector<ui::RenderedContent> rendered;
    rendered.reserve(activities_.size());
    for (const ActivityItemData& activity : activities_) {
        rendered.push_back(renderActivityItem(activity, std::max(1, width)));
    }
    return rendered;
}

}  // namespace activity_feed_example

File: csharp/examples/activity_feed/activity_feed_panel/activity_feed_panel.cs
using Uimd;

internal readonly record struct ActivityItemData(string Timestamp, string EventType, string Message, bool ShowTimestamp);

public sealed class ActivityFeedPanel : ActivityFeedPanelUI
{
    private readonly List<ActivityItemData> activities = new();

    public ActivityFeedPanel()
    {
        SetDynamicChildrenRenderer(RenderActivities);
    }

    public void ClearActivities()
    {
        activities.Clear();
        ClearChildren();
        InvalidateDynamicChildren();
    }

    public void AppendActivity(string timestamp, string eventType, string message, bool showTimestamp)
    {
        activities.Add(new ActivityItemData(timestamp, eventType, message, showTimestamp));
        InvalidateDynamicChildren();
    }

    public void SetAutoScroll(bool enabled)
    {
        ScrollView().SetAutoScroll(enabled);
    }

    public int ActivityCount()
    {
        return activities.Count;
    }

    private List<List<List<TerminalCell>>> RenderActivities(int width)
    {
        List<List<List<TerminalCell>>> rendered = new();
        foreach (ActivityItemData activity in activities)
        {
            rendered.Add(RenderActivityItem(activity, Math.Max(1, width)));
        }
        return rendered;
    }

    private static List<List<TerminalCell>> RenderActivityItem(ActivityItemData item, int width)
    {
        ActivityItemUI view = new();
        view.timestamp.SetText(item.ShowTimestamp ? item.Timestamp : "");
        view.event_type.SetText(item.EventType);
        view.message.SetText(item.Message);

        int sourceHeight = GeneratedSourceHeight(view);
        _ = GeneratedWindowRuntime.RenderGeneratedWindowContent(view, new Size(Math.Max(1, width), sourceHeight));
        int messageRows = WrappedLineCount(item.Message, Math.Max(1, view.message.Frame.Width));
        List<List<TerminalCell>> rendered = GeneratedWindowRuntime.RenderGeneratedWindowContent(
            view,
            new Size(
                Math.Max(1, width),
                CompactContentHeight(view, messageRows)));
        ApplyLeftMarker(rendered, MarkerColor(item.EventType));
        return rendered;
    }

    private static int GeneratedSourceHeight(GeneratedWindowBase view)
    {
        int height = 0;
        foreach (GeneratedLayoutEntry entry in view.GeneratedLayout)
        {
            height = Math.Max(height, entry.SourceCell.Row + entry.SourceCell.Height);
        }
        return Math.Max(1, height);
    }

    private static int CompactContentHeight(GeneratedWindowBase view, int messageRows)
    {
        int height = 1;
        foreach (GeneratedLayoutEntry entry in view.GeneratedLayout)
        {
            if (string.IsNullOrEmpty(entry.Name))
            {
                continue;
            }
            int entryHeight = entry.Name == "message"
                ? Math.Max(1, messageRows)
                : Math.Max(1, entry.CharsSize.Height);
            height = Math.Max(height, entry.Relative.Row + entryHeight);
        }
        return height;
    }

    private static int WrappedLineCount(string text, int width)
    {
        int safeWidth = Math.Max(1, width);
        int count = 0;
        foreach (string line in SplitLines(text))
        {
            count += Math.Max(1, (line.Length + safeWidth - 1) / safeWidth);
        }
        return Math.Max(1, count);
    }

    private static IEnumerable<string> SplitLines(string text)
    {
        string[] lines = text.Split('\n');
        return lines.Length == 0 ? new[] { "" } : lines;
    }

    private static Color MarkerColor(string eventType)
    {
        return eventType switch
        {
            "Task" => new Color("#a78bfa"),
            "Warning" => new Color("#facc15"),
            "Deploy" => new Color("#22c55e"),
            "Note" => new Color("#fb7185"),
            _ => new Color("#38bdf8"),
        };
    }

    private static void ApplyLeftMarker(List<List<TerminalCell>> rendered, Color color)
    {
        foreach (List<TerminalCell> row in rendered)
        {
            if (row.Count == 0)
            {
                continue;
            }
            row[0].Text = "▐";
            row[0].Foreground = color;
        }
    }
}
